### ISSUE TITLE
Uploading Recipies Doesn't Work

### ISSUE THREAD

**Issue by Ryonez** [xthursdayx/ferdi-server-docker#3]
Uploading a recipe does add it to the database, but no files are made:
![image](https://user-images.githubusercontent.com/16157130/66007667-cf7f3000-e50f-11e9-84fe-5a9453c03f2d.png)

This error should probably be marked upstream as well as it breaks the ferdi client recipe list if that custom recipe is meant to be displayed.

Also, this should be made a persistent volume as well.

**Comment by xthursdayx** [comment]
Thanks for this @Ryonez I'll make the recipes directory persistent and update the documentation and will check about the upstream issue. I'll update this issue asap.

**Comment by xthursdayx** [comment]
@Ryonez Can reinstall with the updated image and new unRAID template (which adds a bind path at /path/to/recipes:/usr/src/app/recipes) and let me know if you're getting the same error? Ferdi server has dealt with a similar custom recipe issue, but I believe that it was only with the most recent pre-release beta.

**Comment by xthursdayx** [comment]
@vantezzen do you know if this issue with custom recipes not displaying is related to getferdi/ferdi#70?

**Comment by vantezzen** [comment]
No, they shouldn't bei related

**Comment by xthursdayx** [comment]
@Ryonez Can you describe the process you used to upload your custom recipe?
 
I followed the instructions [here](https://github.com/getferdi/server#creating-and-using-custom-recipes) and got a different error log.

**Comment by vantezzen** [comment]
AdonisJS/ferdi-server expects the `recipes/` (`usr/src/app/recipes/`) to already exist - does the server exist in the Docker image?

**Comment by xthursdayx** [comment]
@vantezzen yeah, the recipes directory is created when the image is created - the Dockerfile just curls and untars the latest Ferdi-server tar.gz from Github (to /usr/src/app/) and then the initialization script installs the AdonisJS app (at /usr/src/app/), migrates the database, and then serves it. The server works fine at this point and all of the correct directories are inside the image, including those that are bind mounted for persistence. 

The issue I've found is that when I create a new custom recipe through `http://[YOUR FERDI-SERVER]/new` I cannot upload the individual recipes files (e.g.index.js, package.json, etc) as described by the getferdi/server README.md, but can only upload a single tar.gz file (despite what the getferdi/server documentation says). If I try to drag and drop the files or select and open them nothing happens; only directories or packaged/compressed files are recognized. After filling in the information on the /new page and uploading a packaged file I get a page that says "new recipe created" but then the AdonisJS server crashes and I have to restart the container. After this I can see that the packaged tar.gz has been added to the persistent /recipes folder, and if I search ferdi:custom in my client I can see the recipe name I created, however nothing happens when I click it to add it.

**Comment by Ryonez** [comment]
@xthursdayx 

Sorry, but I've stopped looking into it for now. The client started randomly having issues logging in and I've need to focus on others things. So I've just closed my things down for now.

**Comment by xthursdayx** [comment]
Strange, it’s working perfectly for me, other than uploading recipes. It would be useful to know what issues you’re facing, so if you decide to try the image out again please let us know what issues you’re facing.

**Comment by vantezzen** [comment]
Normally, you should be able to drag&drop your recipe files into the page - what browser are you using?

The reason for needing to upload the files directly is that:
- Ferdi only recognizes packages recipes that are not packaged inside a subdirectory (i.e. just .tar.gz'ing the whole directory). This is why the packaging is done via AdonisJS: To make sure your recipe gets packaged correctly. This is probably why nothing happens when you upload your own .tar.gz file - the files are not in the right place.
- AdonisJS has some problems when trying to upload a directory to it. Some npm package used by AdonisJS simply crashes the application when there is a directory in the file upload. This is why you'll have to only upload files

Hope we can find a way to improve recipe creation.

**Comment by xthursdayx** [comment]
So, previously I tried creating a new recipe from Firefox and Safari, however, my i think Firefox didn't work because I have it so locked down with ad blocking and privacy settings. I'm not sure what the problem was with Safari, but I just downloaded Google Chrome and tried the whole process again and was able to create a new recipe and upload all the files without any problems. I then tried it again with Safari and everything also worked as expected there. 

Given that custom recipe creation seems to be working fine now I will close this issue, but we can always reopen it if any other issues come up.
